File: android/src/main/kotlin/com/synheart/auth/flutter/SynheartAuthPlugin.kt
package com.synheart.auth.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

/// Flutter plugin that bridges Dart calls to the native SynheartAuth Android SDK.
///
/// In production, this imports and delegates to `com.synheart.auth.SynheartAuth`.
/// The native SDK handles all Android Keystore crypto, storage, and networking.
class SynheartAuthPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.synheart.auth")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> {
                val baseUrl = call.argument<String>("baseUrl")
                    ?: return result.error("INVALID_ARGS", "Missing baseUrl", null)
                com.synheart.auth.SynheartAuth.shared.configure(baseUrl)
                result.success(null)
            }

            "isRegistered" -> {
                val appId = call.argument<String>("appId")
                    ?: return result.error("INVALID_ARGS", "Missing appId", null)
                result.success(com.synheart.auth.SynheartAuth.shared.isRegistered(appId))
            }

            "registerDevice" -> {
                val appId = call.argument<String>("appId")
                    ?: return result.error("INVALID_ARGS", "Missing appId", null)
                scope.launch {
                    try {
                        val reg = com.synheart.auth.SynheartAuth.shared.registerDevice(appId)
                        result.success(
                            mapOf(
                                "status" to reg.status.name.lowercase(),
                                "deviceId" to reg.deviceId
                            )
                        )
                    } catch (e: com.synheart.auth.models.SynheartAuthError) {
                        result.error(errorCode(e), e.message, null)
                    } catch (e: Exception) {
                        result.error("UNKNOWN", e.message, null)
                    }
                }
            }

            "signRequest" -> {
                val appId = call.argument<String>("appId")
                    ?: return result.error("INVALID_ARGS", "Missing appId", null)
                val method = call.argument<String>("method")
                    ?: return result.error("INVALID_ARGS", "Missing method", null)
                val path = call.argument<String>("path")
                    ?: return result.error("INVALID_ARGS", "Missing path", null)
                val bodyBytes = call.argument<ByteArray>("bodyBytes")

                try {
                    val headers = com.synheart.auth.SynheartAuth.shared.signRequest(
                        appId, method, path, bodyBytes
                    )
                    result.success(
                        mapOf(
                            "appId" to headers.appId,
                            "deviceId" to headers.deviceId,
                            "signature" to headers.signature,
                            "timestamp" to headers.timestamp,
                            "nonce" to headers.nonce,
                            "signatureVersion" to headers.signatureVersion
                        )
                    )
                } catch (e: com.synheart.auth.models.SynheartAuthError) {
                    result.error(errorCode(e), e.message, null)
                } catch (e: Exception) {
                    result.error("UNKNOWN", e.message, null)
                }
            }

            "getDeviceId" -> {
                val appId = call.argument<String>("appId")
                    ?: return result.error("INVALID_ARGS", "Missing appId", null)
                result.success(com.synheart.auth.SynheartAuth.shared.getDeviceId(appId))
            }

            "rotateKey" -> {
                val appId = call.argument<String>("appId")
                    ?: return result.error("INVALID_ARGS", "Missing appId", null)
                scope.launch {
                    try {
                        val rot = com.synheart.auth.SynheartAuth.shared.rotateKey(appId)
                        result.success(mapOf("status" to rot.status.name.lowercase()))
                    } catch (e: com.synheart.auth.models.SynheartAuthError) {
                        result.error(errorCode(e), e.message, null)
                    } catch (e: Exception) {
                        result.error("UNKNOWN", e.message, null)
                    }
                }
            }

            "resetDeviceIdentity" -> {
                val appId = call.argument<String>("appId")
                    ?: return result.error("INVALID_ARGS", "Missing appId", null)
                com.synheart.auth.SynheartAuth.shared.resetDeviceIdentity(appId)
                result.success(null)
            }

            "correctClockSkew" -> {
                val serverTimestamp = call.argument<Double>("serverTimestamp")
                    ?: return result.error("INVALID_ARGS", "Missing serverTimestamp", null)
                com.synheart.auth.SynheartAuth.shared.correctClockSkew(serverTimestamp)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun errorCode(e: com.synheart.auth.models.SynheartAuthError): String = when (e) {
        is com.synheart.auth.models.SynheartAuthError.NetworkError -> "NETWORK_ERROR"
        is com.synheart.auth.models.SynheartAuthError.ChallengeExpired -> "CHALLENGE_EXPIRED"
        is com.synheart.auth.models.SynheartAuthError.KeyInvalidated -> "KEY_INVALIDATED"
        is com.synheart.auth.models.SynheartAuthError.ClockSkew -> "CLOCK_SKEW"
        is com.synheart.auth.models.SynheartAuthError.AlreadyRegistered -> "ALREADY_REGISTERED"
        is com.synheart.auth.models.SynheartAuthError.NotRegistered -> "NOT_REGISTERED"
        is com.synheart.auth.models.SynheartAuthError.NotConfigured -> "NOT_CONFIGURED"
        is com.synheart.auth.models.SynheartAuthError.RegistrationInProgress -> "REGISTRATION_IN_PROGRESS"
        is com.synheart.auth.models.SynheartAuthError.ServerError -> e.code
        is com.synheart.auth.models.SynheartAuthError.CryptoError -> "CRYPTO_ERROR"
        is com.synheart.auth.models.SynheartAuthError.StorageError -> "STORAGE_ERROR"
        is com.synheart.auth.models.SynheartAuthError.InvalidStateTransition -> "INVALID_STATE_TRANSITION"
    }
}

File: ios/Classes/SynheartAuthPlugin.swift
import Flutter
import UIKit

/// Flutter plugin that bridges Dart calls to the native SynheartAuth iOS SDK.
///
/// In production, this imports and delegates to the `SynheartAuth` framework.
/// The native SDK handles all Secure Enclave crypto, Keychain storage, and networking.
public class SynheartAuthPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "ai.synheart.auth",
            binaryMessenger: registrar.messenger()
        )
        let instance = SynheartAuthPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "configure":
            guard let baseUrl = args["baseUrl"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing baseUrl", details: nil))
                return
            }
            SynheartAuth.shared.configure(baseUrl: baseUrl)
            result(nil)

        case "isRegistered":
            guard let appId = args["appId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing appId", details: nil))
                return
            }
            result(SynheartAuth.shared.isRegistered(appId: appId))

        case "registerDevice":
            guard let appId = args["appId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing appId", details: nil))
                return
            }
            Task {
                do {
                    let reg = try await SynheartAuth.shared.registerDevice(appId: appId)
                    result([
                        "status": reg.status.rawValue,
                        "deviceId": reg.deviceId as Any,
                    ])
                } catch let error as SynheartAuthError {
                    result(self.flutterError(from: error))
                } catch {
                    result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
                }
            }

        case "signRequest":
            guard let appId = args["appId"] as? String,
                  let method = args["method"] as? String,
                  let path = args["path"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required args", details: nil))
                return
            }
            let bodyBytes = (args["bodyBytes"] as? FlutterStandardTypedData)?.data
            do {
                let headers = try SynheartAuth.shared.signRequest(
                    appId: appId,
                    method: method,
                    path: path,
                    bodyBytes: bodyBytes
                )
                result([
                    "appId": headers.appId,
                    "deviceId": headers.deviceId,
                    "signature": headers.signature,
                    "timestamp": headers.timestamp,
                    "nonce": headers.nonce,
                    "signatureVersion": headers.signatureVersion,
                ])
            } catch let error as SynheartAuthError {
                result(flutterError(from: error))
            } catch {
                result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
            }

        case "getDeviceId":
            guard let appId = args["appId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing appId", details: nil))
                return
            }
            result(SynheartAuth.shared.getDeviceId(appId: appId))

        case "rotateKey":
            guard let appId = args["appId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing appId", details: nil))
                return
            }
            Task {
                do {
                    let rot = try await SynheartAuth.shared.rotateKey(appId: appId)
                    result(["status": rot.status.rawValue])
                } catch let error as SynheartAuthError {
                    result(self.flutterError(from: error))
                } catch {
                    result(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
                }
            }

        case "resetDeviceIdentity":
            guard let appId = args["appId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing appId", details: nil))
                return
            }
            SynheartAuth.shared.resetDeviceIdentity(appId: appId)
            result(nil)

        case "correctClockSkew":
            guard let serverTimestamp = args["serverTimestamp"] as? Double else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing serverTimestamp", details: nil))
                return
            }
            SynheartAuth.shared.correctClockSkew(serverTimestamp: serverTimestamp)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func flutterError(from error: SynheartAuthError) -> FlutterError {
        let (code, message) = switch error {
        case .networkError(let msg): ("NETWORK_ERROR", msg)
        case .challengeExpired: ("CHALLENGE_EXPIRED", "Challenge expired")
        case .attestationUnavailable: ("ATTESTATION_UNAVAILABLE", "Attestation unavailable")
        case .keyInvalidated: ("KEY_INVALIDATED", "Key invalidated")
        case .clockSkew: ("CLOCK_SKEW", "Clock skew detected")
        case .alreadyRegistered: ("ALREADY_REGISTERED", "Already registered")
        case .notRegistered: ("NOT_REGISTERED", "Not registered")
        case .notConfigured: ("NOT_CONFIGURED", "Not configured")
        case .registrationInProgress: ("REGISTRATION_IN_PROGRESS", "Registration in progress")
        case .serverError(let code, let msg): (code, msg)
        case .keychainError(let status): ("KEYCHAIN_ERROR", "Keychain error: \(status)")
        case .cryptoError(let msg): ("CRYPTO_ERROR", msg)
        case .invalidStateTransition(let from, let to): ("INVALID_STATE_TRANSITION", "\(from) → \(to)")
        }
        return FlutterError(code: code, message: message, details: nil)
    }
}
